### PR TITLE
Add table azure_machine_learning_workspace closes #294

### LINKED ISSUE #294: Add table azure_machine_learning_workspace
**References**
https://docs.microsoft.com/en-us/rest/api/azureml/workspaces/list-by-subscription
we need private-endpoint-connections details.
https://docs.microsoft.com/en-us/rest/api/azureml/private-endpoint-connections/list

Note : List call for private endpoint connection details is not supported by SDK till date. https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/services/preview/machinelearningservices/mgmt/2020-02-18-preview/machinelearningservices

Note: This detail is required for NIST SP 800 53 Rev 5 controls.

### PR DESCRIPTION
**Pull request by ParthaI**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/azure_machine_learning_workspace []

PRETEST: tests/azure_machine_learning_workspace

TEST: tests/azure_machine_learning_workspace
Running terraform
data.azurerm_client_config.current: Refreshing state...
data.null_data_source.resource: Refreshing state...
azurerm_resource_group.named_test_resource: Creating...
azurerm_resource_group.named_test_resource: Creation complete after 3s [id=/subscriptions/d46d7416-f95f-4771-bbb5-529d4c76659c/resourceGroups/turbottest41408]
azurerm_application_insights.named_test_resource: Creating...
azurerm_key_vault.named_test_resource: Creating...
azurerm_storage_account.named_test_resource: Creating...
azurerm_application_insights.named_test_resource: Still creating... [10s elapsed]
azurerm_key_vault.named_test_resource: Still creating... [10s elapsed]
azurerm_storage_account.named_test_resource: Still creating... [10s elapsed]
azurerm_application_insights.named_test_resource: Creation complete after 17s [id=/subscriptions/d46d7416-f95f-4771-bbb5-529d4c76659c/resourceGroups/turbottest41408/providers/microsoft.insights/components/turbottest41408]
azurerm_key_vault.named_test_resource: Still creating... [20s elapsed]
azurerm_storage_account.named_test_resource: Still creating... [20s elapsed]
azurerm_key_vault.named_test_resource: Still creating... [30s elapsed]
azurerm_storage_account.named_test_resource: Still creating... [30s elapsed]
azurerm_storage_account.named_test_resource: Creation complete after 38s [id=/subscriptions/d46d7416-f95f-4771-bbb5-529d4c76659c/resourceGroups/turbottest41408/providers/Microsoft.Storage/storageAccounts/turbottest41408]
azurerm_key_vault.named_test_resource: Still creating... [40s elapsed]
azurerm_key_vault.named_test_resource: Still creating... [50s elapsed]
azurerm_key_vault.named_test_resource: Still creating... [1m0s elapsed]
azurerm_key_vault.named_test_resource: Still creating... [1m10s elapsed]
azurerm_key_vault.named_test_resource: Still creating... [1m20s elapsed]
azurerm_key_vault.named_test_resource: Still creating... [1m30s elapsed]
azurerm_key_vault.named_test_resource: Still creating... [1m40s elapsed]
azurerm_key_vault.named_test_resource: Still creating... [1m50s elapsed]
azurerm_key_vault.named_test_resource: Still creating... [2m0s elapsed]
azurerm_key_vault.named_test_resource: Still creating... [2m10s elapsed]
azurerm_key_vault.named_test_resource: Still creating... [2m20s elapsed]
azurerm_key_vault.named_test_resource: Still creating... [2m30s elapsed]
azurerm_key_vault.named_test_resource: Creation complete after 2m36s [id=/subscriptions/d46d7416-f95f-4771-bbb5-529d4c76659c/resourceGroups/turbottest41408/providers/Microsoft.KeyVault/vaults/turbottest41408]
azurerm_machine_learning_workspace.named_test_resource: Creating...
azurerm_machine_learning_workspace.named_test_resource: Still creating... [10s elapsed]
azurerm_machine_learning_workspace.named_test_resource: Still creating... [20s elapsed]
azurerm_machine_learning_workspace.named_test_resource: Still creating... [30s elapsed]
azurerm_machine_learning_workspace.named_test_resource: Still creating... [40s elapsed]
azurerm_machine_learning_workspace.named_test_resource: Still creating... [50s elapsed]
azurerm_machine_learning_workspace.named_test_resource: Still creating... [1m0s elapsed]
azurerm_machine_learning_workspace.named_test_resource: Still creating... [1m10s elapsed]
azurerm_machine_learning_workspace.named_test_resource: Creation complete after 1m14s [id=/subscriptions/d46d7416-f95f-4771-bbb5-529d4c76659c/resourceGroups/turbottest41408/providers/Microsoft.MachineLearningServices/workspaces/turbottest41408]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 5 added, 0 changed, 0 destroyed.

Outputs:

location = eastus
resource_aka = azure:///subscriptions/d46d7416-f95f-4771-bbb5-529d4c76659c/resourceGroups/turbottest41408/providers/Microsoft.MachineLearningServices/workspaces/turbottest41408
resource_aka_lower = azure:///subscriptions/d46d7416-f95f-4771-bbb5-529d4c76659c/resourcegroups/turbottest41408/providers/microsoft.machinelearningservices/workspaces/turbottest41408
resource_id = /subscriptions/d46d7416-f95f-4771-bbb5-529d4c76659c/resourceGroups/turbottest41408/providers/Microsoft.MachineLearningServices/workspaces/turbottest41408
resource_name = turbottest41408
subscription_id = d46d7416-f95f-4771-bbb5-529d4c76659c

Running SQL query: test-get-query.sql
[
  {
    "id": "/subscriptions/d46d7416-f95f-4771-bbb5-529d4c76659c/resourceGroups/turbottest41408/providers/Microsoft.MachineLearningServices/workspaces/turbottest41408",
    "name": "turbottest41408",
    "region": "eastus",
    "resource_group": "turbottest41408",
    "subscription_id": "d46d7416-f95f-4771-bbb5-529d4c76659c",
    "tags": {
      "name": "turbottest41408"
    },
    "type": "Microsoft.MachineLearningServices/workspaces"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "id": "/subscriptions/d46d7416-f95f-4771-bbb5-529d4c76659c/resourceGroups/turbottest41408/providers/Microsoft.MachineLearningServices/workspaces/turbottest41408",
    "name": "turbottest41408",
    "type": "Microsoft.MachineLearningServices/workspaces"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "azure:///subscriptions/d46d7416-f95f-4771-bbb5-529d4c76659c/resourceGroups/turbottest41408/providers/Microsoft.MachineLearningServices/workspaces/turbottest41408",
      "azure:///subscriptions/d46d7416-f95f-4771-bbb5-529d4c76659c/resourcegroups/turbottest41408/providers/microsoft.machinelearningservices/workspaces/turbottest41408"
    ],
    "name": "turbottest41408",
    "tags": {
      "name": "turbottest41408"
    },
    "title": "turbottest41408"
  }
]
✔ PASSED

POSTTEST: tests/azure_machine_learning_workspace

TEARDOWN: tests/azure_machine_learning_workspace

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
### Basic info

```sql
select
  name,
  id,
  type,
  provisioning_state
from
  azure_machine_learning_workspace;
```

```
+------+------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------+----------
| name | id                                                                                                                                       | type                                         | provision
+------+------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------+----------
| test | /subscriptions/d46d7416-f95f-4771-bbb5-529d4c76659c/resourceGroups/turbot_rg/providers/Microsoft.MachineLearningServices/workspaces/test | Microsoft.MachineLearningServices/workspaces | Succeeded
+------+------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------+---------
```

### List system assigned identity type workspace

```sql
select
  name,
  id,
  type,
  identity ->> 'type' as identity_type
from
  azure_machine_learning_workspace
where
  identity ->> 'type' = 'SystemAssigned';
```

```
+------+------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------+----------
| name | id                                                                                                                                       | type                                         | identity_
+------+------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------+----------
| test | /subscriptions/d46d7416-f95f-4771-bbb5-529d4c76659c/resourceGroups/turbot_rg/providers/Microsoft.MachineLearningServices/workspaces/test | Microsoft.MachineLearningServices/workspaces | SystemAss
+------+------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------+----------
```

### List key vault used by workspaces with soft deletion enabled

```sql
select
  m.name as workspace_name,
  m.id as workspace_id,
  v.soft_delete_enabled
from
  azure_machine_learning_workspace as m,
  azure_key_vault as v
where
  lower(m.key_vault) = lower(v.id) and v.soft_delete_enabled;
```

```
+----------------+------------------------------------------------------------------------------------------------------------------------------------------+---------------------+
| workspace_name | workspace_id                                                                                                                             | soft_delete_enabled |
+----------------+------------------------------------------------------------------------------------------------------------------------------------------+---------------------+
| test           | /subscriptions/d46d7416-f95f-4771-bbb5-529d4c76659c/resourceGroups/turbot_rg/providers/Microsoft.MachineLearningServices/workspaces/test | true                |
+----------------+------------------------------------------------------------------------------------------------------------------------------------------+---------------------+
```

</details>

### CODE REVIEW

File: azure-test/tests/azure_machine_learning_workspace/variables.tf [hunk 1]
```diff
@@ -0,0 +1,103 @@
+
```
Comment (bigdatasourav) [inline azure-test/tests/azure_machine_learning_workspace/variables.tf:1]:
```suggestion
```

File: azure/table_azure_machine_learning_workspace.go [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+	"github.com/turbot/steampipe-plugin-sdk/plugin"
+)
+
+//// TABLE DEFINITION
+
+func tableAzureMachineLearningWorkspace(_ context.Context) *plugin.Table {
+	return &plugin.Table{
+		Name:        "azure_machine_learning_workspace",
+		Description: "Azure Machine Learning Worspace",
```
Comment (bigdatasourav) [inline azure/table_azure_machine_learning_workspace.go:19]:
```suggestion
		Description: "Azure Machine Learning Workspace",
```

File: azure/table_azure_machine_learning_workspace.go [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+				Type:        proto.ColumnType_JSON,
+			},
+			{
+				Name:        "akas",
+				Description: ColumnDescriptionAkas,
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("ID").Transform(idToAkas),
+			},
+
+			// Azure standard column
```
Comment (bigdatasourav) [inline azure/table_azure_machine_learning_workspace.go:148]:
```suggestion
			// Azure standard columns
```

File: azure/table_azure_machine_learning_workspace.go [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+	plugin.Logger(ctx).Trace("getDataFactory")
+
+	// Create session
+	session, err := GetNewSession(ctx, d, "MANAGEMENT")
+	if err != nil {
+		return nil, err
+	}
+	subscriptionID := session.SubscriptionID
+
+	worspaceClient := machinelearningservices.NewWorkspacesClient(subscriptionID)
```
Comment (bigdatasourav) [inline azure/table_azure_machine_learning_workspace.go:278]:
```suggestion
	workspaceClient := machinelearningservices.NewWorkspacesClient(subscriptionID)
```

File: azure/table_azure_machine_learning_workspace.go [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+	worspaceClient.Authorizer = session.Authorizer
```
Comment (bigdatasourav) [inline azure/table_azure_machine_learning_workspace.go:279]:
```suggestion
	workspaceClient.Authorizer = session.Authorizer
```

File: azure/table_azure_machine_learning_workspace.go [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+
+	name := d.KeyColumnQuals["name"].GetStringValue()
+	resourceGroup := d.KeyColumnQuals["resource_group"].GetStringValue()
+
+	// Return nil, if no input provide
+	if name == "" || resourceGroup == "" {
+		return nil, nil
+	}
+
+	workspace, err := worspaceClient.Get(ctx, resourceGroup, name)
```
Comment (bigdatasourav) [inline azure/table_azure_machine_learning_workspace.go:289]:
```suggestion
	workspace, err := workspaceClient.Get(ctx, resourceGroup, name)
```

**Review by bigdatasourav — COMMENTED** [review]
@ParthaI as discussed, need to check if we require a custom structure or not.

File: azure/table_azure_machine_learning_workspace.go [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+		return nil, err
+	}
+	subscriptionID := session.SubscriptionID
+
+	worspaceClient := machinelearningservices.NewWorkspacesClient(subscriptionID)
+	worspaceClient.Authorizer = session.Authorizer
+
+	result, err := worspaceClient.ListBySubscription(ctx, "")
+	if err != nil {
+		return nil, err
```
Comment (bigdatasourav) [inline azure/table_azure_machine_learning_workspace.go:207]: log error

File: azure/table_azure_machine_learning_workspace.go [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+	resourceGroup := d.KeyColumnQuals["resource_group"].GetStringValue()
+
+	// Return nil, if no input provide
+	if name == "" || resourceGroup == "" {
+		return nil, nil
+	}
+
+	workspace, err := worspaceClient.Get(ctx, resourceGroup, name)
+	if err != nil {
+		return nil, err
```
Comment (bigdatasourav) [inline azure/table_azure_machine_learning_workspace.go:291]: Add logger to log error

**Comment by ParthaI** [comment]
> @ParthaI as discussed, need to check if we require a custom structure or not.

We need custom structure I checked we are getting few attributes value as null like `friendly_name`, `provisioning_state`, `creation_time`, `workspace_id` etc..

File: azure/table_azure_machine_learning_workspace.go [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "type",
+				Description: "The resource type.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "diagnostic_settings",
+				Description: "A list of active diagnostic settings for the n workspace.",
```
Comment (bigdatasourav) [inline azure/table_azure_machine_learning_workspace.go:141]:
```suggestion
				Description: "A list of active diagnostic settings for the ML workspace.",
```

**Review by bigdatasourav — APPROVED** [review]
LGTM

**Pull request merged**